src/nvim/drawscreen.c

```diff
@@ -733,28 +734,53 @@ void show_cursor_info(bool always)
     return;
   }
 
-  win_check_ns_hl(curwin);
-  if ((*p_stl != NUL || *curwin->w_p_stl != NUL)
-      && (curwin->w_status_height || global_stl_height())) {
-    redraw_custom_statusline(curwin);
-  } else {
-    win_redr_ruler(curwin, always);
-  }
-  if (*p_wbr != NUL || *curwin->w_p_wbr != NUL) {
-    win_redr_winbar(curwin);
-  }
+  int state = get_real_state();
+  int empty_line = (State & MODE_INSERT) == 0
+                   && *ml_get_buf(curwin->w_buffer, curwin->w_cursor.lnum, false) == NUL;
+
+  // Only draw when something changed.
+  validate_virtcol_win(curwin);
+  if (always
+      || curwin->w_cursor.lnum != curwin->w_stl_cursor.lnum
+      || curwin->w_cursor.col != curwin->w_stl_cursor.col
+      || curwin->w_virtcol != curwin->w_stl_virtcol
+      || curwin->w_cursor.coladd != curwin->w_stl_cursor.coladd
+      || curwin->w_topline != curwin->w_stl_topline
+      || curwin->w_buffer->b_ml.ml_line_count != curwin->w_stl_line_count
+      || curwin->w_topfill != curwin->w_stl_topfill
+      || empty_line != curwin->w_stl_empty
+      || state != curwin->w_stl_state) {
+    win_check_ns_hl(curwin);
+    if ((*p_stl != NUL || *curwin->w_p_stl != NUL)
+        && (curwin->w_status_height || global_stl_height())) {
+      redraw_custom_statusline(curwin);
+    } else {
+      win_redr_ruler(curwin);
+    }
+    if (*p_wbr != NUL || *curwin->w_p_wbr != NUL) {
+      win_redr_winbar(curwin);
+    }
 
-  if (need_maketitle
-      || (p_icon && (stl_syntax & STL_IN_ICON))
-      || (p_title && (stl_syntax & STL_IN_TITLE))) {
-    maketitle();
-  }
+    if (need_maketitle
```

Comment:
Why are this and the `draw_tabline()` below conditional on `need_maketitle/redraw_tabline`? Either we want to support showing the cursor(now mode) information in the title/tabline or we don't. If we do they should be unconditionally called here I think, `set tal=%{mode()}` will not show the mode changes for example. Them being present in `show_cursor_info()` seems to indicate that we want to support this.

Reply:
I agree it is weird, but let's consider that in a follow up PR (as I said, I want to refactor a lot of this code anyway)

Reply:
although maybe the alternate `stl_syntax & STL_IN_TITLE)` condition handles this? too many magic variables..

Reply:
Yeah for the title maybe. But like I said set `tal=%{mode()}` will not show mode changes due to the guard on `redraw_tabline` here.

Another weird thing: there is a `maketitle()` in `normal_redraw()` right before `show_cursor_info()` which will be redundant depending on what we do with this.